base/logging_new.go

```diff
@@ -0,0 +1,137 @@
+package base
+
+import (
+	"log"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	// RFC3339 + 3 digit milli
+	timestampFormat = "2006-01-02T15:04:05.000Z07:00"
+)
+
+var (
+	ErrInvalidLogFilePath   = errors.New("invalid LogFilePath")
+	ErrInvalidLoggingMaxAge = errors.New("invalid MaxAge")
+
+	// We'll initilise a default consoleLogger so we can still log stuff before/during parsing logging configs.
+	defaultLogLevel = LEVEL_INFO
+	defaultLogKey   = KEY_ALL
+	consoleLogger   = LogConsoleConfig{
+		LogLevel:     &defaultLogLevel,
+		logKey:       &defaultLogKey,
+		logger:       log.New(os.Stderr, "", 0),
+		ColorEnabled: true,
+	}
+
+	debugLogger, infoLogger, warnLogger, errorLogger LogFileConfig
+)
+
+// Errorf logs the given formatted string and args to the error log level and given log key.
+func Errorf(logKey LogKey, format string, args ...interface{}) {
+	logTo(LEVEL_ERROR, logKey, format, args...)
+}
+
+// Warnf logs the given formatted string and args to the warn log level and given log key.
+func Warnf(logKey LogKey, format string, args ...interface{}) {
+	logTo(LEVEL_WARN, logKey, format, args...)
+}
+
+// Infof logs the given formatted string and args to the info log level and given log key.
+func Infof(logKey LogKey, format string, args ...interface{}) {
+	logTo(LEVEL_INFO, logKey, format, args...)
+}
+
+// Debugf logs the given formatted string and args to the debug log level with an optional log key.
+func Debugf(logKey LogKey, format string, args ...interface{}) {
+	logTo(LEVEL_DEBUG, logKey, format, args...)
+}
+
+func logTo(logLevel LogLevel, logKey LogKey, format string, args ...interface{}) {
+	// exit early for things we know won't get logged
+	if logKey <= KEY_NONE || logLevel <= LEVEL_NONE {
+		return
+	}
+
```

Comment:
It feels like we should be doing the log level check before the string formatting (add prefixes, redact) - otherwise it feels like we're doing a lot of unnecessary work for debug logging in the default scenario.

Maybe push the shouldLog check up to Logf/Warnf etc?

Reply:
Yeah agreed. I'm doing the shouldLog checks up front now, and exiting early if they're all false before continuing with formatting/redacting, etc.